Adds ability to watch files and exit gracefully if they change

JoshVanL opened the pull request:
```release-note
Adds options to enable watching files that cause a graceful exit when modified
```

In favor of #88 

/hold
/assign
fixes #16
/milestone v0.2

JoshVanL commented:
Should be good to have a look now
/unassign
/assign @simonswine 
/hold cancel
/test pull-kube-oidc-proxy-e2e-v1-14
/test pull-kube-oidc-proxy-e2e-v1-13
/test pull-kube-oidc-proxy-e2e-v1-12
/test pull-kube-oidc-proxy-e2e-v1-11

JoshVanL commented:
/retest

simonswine commented on test/e2e/suite/cases/watch/watch.go:
```diff
@@ -0,0 +1,155 @@
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Waiting for proxy to pick up change")
+		time.Sleep(time.Second * 15)
```
Not a big fan of sleep in testing. Can we use a retry mechanism?

simonswine commented on test/e2e/suite/cases/watch/watch.go:
```diff
@@ -0,0 +1,155 @@
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Waiting for proxy to pick up change")
+		time.Sleep(time.Second * 15)
```
Not a big fan of sleep in testing. Can we use a retry mechanism?

JoshVanL commented:
@simonswine 
I have updated the docs to make it clear it will terminate the proxy on watch change and mentioned the HPA issue.
I think the CLI flag is probably clear enough the exit behavior.

`Comma-separated list of files to watch for changes. If a change is detected then send self a SIGHUP signal and exit gracefully.` 

I've updated the tests and will re-assign once they've passed.

JoshVanL commented:
This functionality should be checked whether this has been included, and if not, enabled using apiserver module

/close

jescarri commented:
What's the status on this PR? is this a feature that's not going to be included in future releases?

Thanks!

JoshVanL commented:
Hey @jescarri,

In theory, now that we have updated dependencies on k8s to 1.18, this should be included as default. I haven't tested this however, so if it isn't the case than I will do some investigating to make sure it's there.
